applications/RomApplication/costum_strategies/rom_builder_and_solver.h

```diff
@@ -0,0 +1,811 @@
+		for (auto it = dof_global_set.begin(); it != dof_global_set.end(); it++)
+		{
+			Doftemp.push_back(it->get());
+		}
+		Doftemp.Sort();
+
+		BaseType::mDofSet = Doftemp;
+
+		//Throws an exception if there are no Degrees Of Freedom involved in the analysis
+		KRATOS_ERROR_IF(BaseType::mDofSet.size() == 0) << "No degrees of freedom!" << std::endl;
+
+		KRATOS_INFO_IF("ROMBuilderAndSolver", (this->GetEchoLevel() > 2)) << "Number of degrees of freedom:" << BaseType::mDofSet.size() << std::endl;
+
+		BaseType::mDofSetIsInitialized = true;
+
+		KRATOS_INFO_IF("ROMBuilderAndSolver", (this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished setting up the dofs" << std::endl;
+
+		KRATOS_INFO_IF("ROMBuilderAndSolver", (this->GetEchoLevel() > 2)) << "End of setup dof set\n" << std::endl;
+
+
+#ifdef KRATOS_DEBUG
+		// If reactions are to be calculated, we check if all the dofs have reactions defined
+		// This is tobe done only in debug mode
+		if (BaseType::GetCalculateReactionsFlag()) {
+			for (auto dof_iterator = BaseType::mDofSet.begin(); dof_iterator != BaseType::mDofSet.end(); ++dof_iterator) {
+				KRATOS_ERROR_IF_NOT(dof_iterator->HasReaction()) << "Reaction variable not set for the following : " << std::endl
+					<< "Node : " << dof_iterator->Id() << std::endl
+					<< "Dof : " << (*dof_iterator) << std::endl << "Not possible to calculate reactions." << std::endl;
+			}
+		}
+#endif
+		{
+			mDofList.clear();
+			for (auto& node : rModelPart.Nodes())
+			{
+				for (const std::string& var_name : mNodalVariablesNames)
+				{
+					if (KratosComponents< Variable<double> >::Has(var_name)) //case of double variable
+					{
+						auto pdof = node.pGetDof(KratosComponents< Variable<double> >::Get(var_name));
+						mDofList.push_back(pdof);
+					}
+					else if (KratosComponents< VariableComponent< VectorComponentAdaptor<array_1d<double, 3> > > >::Has(var_name)) //case of component variable
+					{
+						typedef VariableComponent< VectorComponentAdaptor<array_1d<double, 3> > > component_type;
+						component_type var_component = KratosComponents< component_type >::Get(var_name);
+						auto pdof = node.pGetDof(var_component.GetSourceVariable());
+						mDofList.push_back(pdof);
+					}
+				}
+			}
+		}
+
+
+		KRATOS_CATCH("");
+	}
+
+    /**
+            organises the dofset in order to speed up the building phase
+     */
+    virtual void SetUpSystem(
+        ModelPart& r_model_part
+    ) override
+    {
+		//int free_id = 0;
+		BaseType::mEquationSystemSize = BaseType::mDofSet.size();
+		int ndofs = static_cast<int>(BaseType::mDofSet.size());
+
+#pragma omp parallel for firstprivate(ndofs)
+		for (int i = 0; i < static_cast<int>(ndofs); i++) {
+			typename DofsArrayType::iterator dof_iterator = BaseType::mDofSet.begin() + i;
+			dof_iterator->SetEquationId(i);
+
+		}
+    }
+
+    Vector ProjectToReducedBasis(const TSystemVectorType& rX, ModelPart::NodesContainerType& rNodes)
```

Comment:
Why do you return a vector? What we normally do in these classes is to pass a reference to be filled / modified.